app/services/course/assessment/submission/statistics_download_service.rb

```diff
@@ -0,0 +1,143 @@
+# frozen_string_literal: true
+require 'csv'
+class Course::Assessment::Submission::StatisticsDownloadService
+  include ApplicationFormattersHelper
+
+  class << self
+    # Downloads the submissions statistics and zip it.
+    #
+    # @param [CourseUser] course_user The course user downloading the statistics.
+    # @param [Array<Course::Assessment::Submission>] submissions The submissions to download
+    # statistics
+    # @return [String] The path to the zip file.
+    def download_and_zip(course_user, submissions)
+      service = new(course_user, submissions)
+      ActsAsTenant.without_tenant do
+        service.send(:download_to_base_dir)
+      end
+      service.send(:zip_base_dir)
+    end
+  end
+
+  private
+
+  def initialize(course_user, submissions)
+    @course_user = course_user
+    @submissions = submissions
```

Comment:
According to weiqingtoh's comment, I moved query into the controller. The problem of ```ApplicationJob``` is it cannot receive Association as the input parameter. Hence, all the eager load in the controller won't help in the service. 
Hence, I think we need to move the query back to the service?

Reply:
May I ask if ```graded_at``` will benefit from the eager load of answers?  Since ```graded_at``` does not use Associations, and it use plain sql?

Reply:
> According to weiqingtoh's comment, I moved query into the controller. The problem of ```ApplicationJob``` is it cannot receive Association as the input parameter. Hence, all the eager load in the controller won't help in the service. 
Hence, I think we need to move the query back to the service? 

You can just `pluck` the submission_ids, and send that as a param into the service, rather than sending the entire set of submission objects. Then in the service, do the necessary loading eager loading for the service - so when the service is updated next time, the engineer just needs to modify this in the `initialize` method. 

> May I ask if ```graded_at``` will benefit from the eager load of answers?  Since ```graded_at``` does not use Associations, and it use plain sql? 

It depends on the implementation. For our case, `calculated_attributes` gets `graded_at` through a pure SQL injected as a sub-query, so including the association does not provide any benefit (since it is still done in 1 query). 

If for some reason we use a ruby method to calculate `graded_at`, each `submission` would require the `answers` association, and yes it might be better to eager load then. 

---

Hope this clarifies @huangjosh

Reply:
Thanks @weiqingtoh . This is very helpful.
I think I will follow you suggestion to improve this PR.